### PR TITLE
[WIP] :sparkles: Initial drop of software

### PR DESCRIPTION
**Pull request by FluffyFTW**
- Added power settings
- Added GPIO configs
- Added output pin constructor

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -32,27 +32,26 @@ class libhal_stm32f4_conan(ConanFile):
...
+    topics = ("arm", "microcontroller", "stm32f4",)
     settings = "compiler", "build_type", "os", "arch"
     exports_sources = ("include/*", "linker_scripts/*", "tests/*", "LICENSE",
                        "CMakeLists.txt", "src/*")
     generators = "CMakeToolchain", "CMakeDeps", "VirtualBuildEnv"
 
     options = {
-        "platform": [
-            "profile1",
-            "profile2",
```
Comment (kammce) [inline conanfile.py:44]: You should put stm32f411re here

File: conanfile.py [hunk 2]
```diff
@@ -32,27 +32,26 @@ class libhal_stm32f4_conan(ConanFile):
...
 
+    def package_id(self):
+        if self.info.options.get_safe("platform"):
+            del self.info.options.platform
+
     @property
     def _use_linker_script(self):
-        return (self.options.platform == "profile1" or
+        return (self.options.platform == "stm32f4" or
                 self.options.platform == "profile2")
```
Comment (kammce) [inline conanfile.py]: Consider removing the other profile?

File: include/libhal-stm32f4/output_pin.hpp [hunk 3]
```diff
@@ -20,13 +20,28 @@ namespace hal::stm32f4 {
...
+  /**
+   * @brief Get the output pin object
+   *
+   * @param p_port - selects pin port to use
+   * @param p_pin - selects which pin within the port to use
+   * @param p_settings - initial pin settings
+   * @return result<output_pin> - reference to the statically allocated output
+   * pin
+   */
+  static result<output_pin> get(std::char p_port,
```
Comment (kammce) [inline include/libhal-stm32f4/output_pin.hpp]:
```suggestion
  static result<output_pin> get(char p_port,
```

File: include/libhal-stm32f4/output_pin.hpp [hunk 4]
```diff
@@ -20,13 +20,28 @@ namespace hal::stm32f4 {
...
+                                std::uint8_t p_pin,
+                                output_pin::settings p_settings = {});
+
   static hal::result<output_pin> create();
 
 private:
-  // Add constructor
+  output_pin(std::char p_port, std::uint8_t p_pin);
```
Comment (kammce) [inline include/libhal-stm32f4/output_pin.hpp]:
```suggestion
  output_pin(char p_port, std::uint8_t p_pin);
```

File: src/gpio_reg.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+inline constexpr intptr_t ahb_base = 0x40020000UL;
+
+enum class gpio_reg{
+  a = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x00000),
+  b = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x00400),
+  c = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x00800),
+  d = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x00C00),
+  e = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x01000),
+  h = reinterpret_cast<stm32f4_gpio_t*>(ahb_base + 0x01C00)
+};
```
Comment (kammce) [inline src/gpio_reg.hpp]: Thats an interesting way to do this. Feel free to resolve this. This is just a comment.

**Review by kammce — COMMENTED** [review]
Looking pretty good for the start, keep it up :)

File: conanfile.py (hunk 1 above)
Comment (FluffyFTW) [reply conanfile.py:44]: Just curious, but when refering to both lpc40 and stm32f1 the platforms are both ["ANY"], why would the stm32f4 be different?

Comment (kammce) [reply conanfile.py:44]: Nah, you're right, keep it as "any".

**Pull request closed without merging**